optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,480 @@
+                        Joiners.equal(DayOffRequest::getShiftDate, ShiftAssignment::getShiftDate))
+                .penalize("org.optaplanner.examples.nurserostering.solver", "dayOffRequest", HardSoftScore.ONE_SOFT,
+                        (dayOffRequest, shiftAssignment) -> dayOffRequest.getWeight());
+    }
+
+    Constraint dayOnRequest(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(DayOnRequest.class)
+                .ifNotExists(ShiftAssignment.class, Joiners.equal(DayOnRequest::getEmployee, ShiftAssignment::getEmployee),
+                        Joiners.equal(DayOnRequest::getShiftDate, ShiftAssignment::getShiftDate))
+                .penalize("org.optaplanner.examples.nurserostering.solver", "dayOnRequest", HardSoftScore.ONE_SOFT,
+                        DayOnRequest::getWeight);
+    }
+
+    // Requested shift on/off
+    Constraint shiftOffRequest(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(ShiftOffRequest.class)
+                .join(ShiftAssignment.class, Joiners.equal(ShiftOffRequest::getEmployee, ShiftAssignment::getEmployee),
+                        Joiners.equal(ShiftOffRequest::getShift, ShiftAssignment::getShift))
+                .penalize("org.optaplanner.examples.nurserostering.solver", "shiftOffRequest", HardSoftScore.ONE_SOFT,
+                        (shiftOffRequest, shiftAssignment) -> shiftOffRequest.getWeight());
+    }
+
+    Constraint shiftOnRequest(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(ShiftOnRequest.class)
+                .ifNotExists(ShiftAssignment.class, Joiners.equal(ShiftOnRequest::getEmployee, ShiftAssignment::getEmployee),
+                        Joiners.equal(ShiftOnRequest::getShift, ShiftAssignment::getShift))
+                .penalize("org.optaplanner.examples.nurserostering.solver", "shiftOnRequest", HardSoftScore.ONE_SOFT,
+                        ShiftOnRequest::getWeight);
+    }
+
+    // Alternative skill
+    Constraint alternativeSkill(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BooleanContractLine.class)
+                .filter(booleanContractLine -> booleanContractLine.getContractLineType()
+                        .equals(ContractLineType.ALTERNATIVE_SKILL_CATEGORY))
+                .join(ShiftAssignment.class, Joiners.equal(BooleanContractLine::getContract, ShiftAssignment::getContract))
+                .join(ShiftTypeSkillRequirement.class,
+                        Joiners.equal((contract, sa) -> sa.getShiftType(), ShiftTypeSkillRequirement::getShiftType))
+                .ifNotExists(SkillProficiency.class,
+                        Joiners.equal((contract, sa, stsr) -> sa.getEmployee(), SkillProficiency::getEmployee),
+                        Joiners.equal((contract, sa, stsr) -> stsr.getSkill(), SkillProficiency::getSkill))
```

Comment:
We generally don't use abbreviations like `sa`and `stsr`.
I understand the need to shorten these long names, but `assignment` and `skillRequirement` should be OK.